File: EP/parser.js
var fs = require('fs');
var net = require('net');
var Utils = require('./utils')


VERTEX_NAME_CODE_MAPPING = require('./mapper.json')


var Class = function(methods) {
    var klass = function() {
        this.initialize.apply(this, arguments);
    };

    for (var property in methods) {
       klass.prototype[property] = methods[property];
    }

    if (!klass.prototype.initialize) klass.prototype.initialize = function(){};

    return klass;
};


var Parser = Class({
    initialize: function(){
        this.active = null
        this.__segments = []
    },

    deactivate: function(){
        that = this
        that.__segments.forEach(function(val, idx){
            if (val['st'] === 'FUTURE'){
                that.__segments[idx]['st'] = 'INACT'
            }
        })

    },

    arrival: function(sdt){
        if (sdt === undefined){
            return this.__segments[this.active]['a_arr']
        } else {
            this.__segments[this.active]['a_arr'] = sdt
        }
    },

    add_nullseg: function(){
        var segment = {
            'src': null,
            'dst': null,
            'conn': null,
            'p_arr': null,
            'a_arr': null,
            'p_dep': null,
            'a_dep': null,
            'cst': null,
            'st': 'REACHED',
            'rmk': [],
            'idx': 0,
            'par': null,
            'sol': null,
            'pdd': null,
        }
        this.__segments.push(segment)
        this.active = 0
    },

    add_segment: function(subgraph, kwargs){

        if (subgraph === true){
            if (this.__segments.length === 0){
                this.add_nullseg()
            }
        }
        var segment = {
            'src': kwargs['src'],
            'dst': kwargs['dst'],
            'conn': kwargs['conn'],
            'p_arr': kwargs['p_arr'],
            'p_dep': kwargs['p_dep'],
            'cst': kwargs['cst'],
            'a_arr': kwargs['a_arr'] ||  null,
            'a_dep': kwargs['a_dep'] || null,
            'st':  kwargs['st'] || 'FUTURE',
            'rmk': kwargs['rmk'] || [],
        }

        for (var key in kwargs){

            if (['src', 'dst', 'conn', 'p_arr', 'p_dep', 'cst', 'a_arr', 'a_dep', 'st', 'rmk', 'idx', 'par'].indexOf(key) < 0){
                segment[key] = kwargs[key]
            }
        }

        segment['idx'] = kwargs['idx'] || this.__segments.length

        if (subgraph != true){
            var par = kwargs['par']
            if (par === undefined){
                par = this.__segments.length - 1
            }
            segment['par'] = par
        } else{
            segment['par'] = this.active
            segment['st'] = 'ACTIVE'
        }

        if (segment['st'] === 'ACTIVE') {
            this.active = segment['idx']
        }

        this.__segments.push(segment)
        return this.__segments.length
    },

    add_segments: function(segments, novi){
        var self = this
        if (novi === undefined){
            novi = false
        }
        segments.forEach(function(segment){
            try{
                self.add_segment(novi, segment)
            } catch(err){
                console.error('ERROR: '+err)
            }
            novi = false
        })

    },
    mark_inbound: function(scan_datetime, rmk, fail, arrived){
        var active  = this.active

        if (arrived){
            this.__segments[this.active]['a_arr'] = scan_datetime
        }
        if (rmk !== undefined) {
            this.__segments[this.active]['rmk'].push(rmk)
        }

        if (fail) {
            this.__segments[active]['st'] = 'FAIL'
            this.deactivate()
            var active = this.__segments[active]['par']
            this.active = active
        }
    },

    mark_outbound: function(scan_datetime, rmk, fail, departed){
        if (departed){
            this.__segments[this.active]['a_dep'] = scan_datetime
        }

        if (rmk !== undefined) {
            this.__segments[this.active]['rmk'].push(rmk)
        }

        if (fail) {
            this.__segments[this.active]['st'] = 'FAIL'
            this.deactivate()
            var active = this.__segments[this.active]['par']
            this.active = active
        }
        else {
            this.__segments[this.active]['st'] = 'REACHED'
            var active = this.lookup({
                'par': this.__segments[this.active]['idx'],
                'st': 'FUTURE'
            })
            this.active = active
            this.__segments[this.active]['st'] = 'ACTIVE'
        }
    },

    make_new: function(connection, intermediary, source){
        if (source === undefined || source === null){
            source = this.active
        }
        var segment = {}

        for (var key in this.__segments[this.active]){
            segment[key] = this.__segments[this.active][key]
        }
        segment['conn'] = connection
        segment['dst'] = intermediary
        segment['st'] = 'REACHED'
        segment['idx'] = this.__segments.length
        this.__segments.push(segment)
        this.active = segment['idx']
    },

    make_new_blank: function(src, dst, cid, sdt){
        var segment = {
            'src': src,
            'dst': dst,
            'conn': cid,
            'p_arr': null,
            'a_arr': null,
            'p_dep': null,
            'a_dep': sdt,
            'cst': this.__segments[this.active]['cst'],
            'st': 'REACHED',
            'rmk': [],
            'par': this.active,
            'sol': this.__segments[this.active]['sol'],
            'pdd': this.__segments[this.active]['pdd']
        }
        segment['idx'] = this.__segments.length
        this.__segments.push(segment)
        var active = segment['idx']
        this.active = active
    },

    mark_termination: function(msg){
        this.__segments[this.active]['rmk'].push(msg)
        this.__segments[this.active]['st'] = 'FAIL'
    },

    parse_inbound: function(location, scan_datetime){
        if (this.active == null){
            var err = 'ERROR (parse_inbound): No active nodes found'
            console.error(err)
            throw err
        }
        a_seg = this.__segments[this.active]

        if (a_seg['src'] === location){
            if (a_seg['a_arr'] !== null){
                console.log('Duplicate inbound')
                return true
            }
            if (a_seg['p_arr'] >= scan_datetime){
                this.mark_inbound(scan_datetime, undefined, false, true)
                return true
            } else if (scan_datetime < a_seg['p_dep']){
                this.mark_inbound(scan_datetime, 'WARN_LATE_ARRIVAL', false, true)
                return true
            } else {
                this.mark_inbound(
                    scan_datetime, 'FAIL_LATE_ARRIVAL', true, true)
            }
        }
        else{
            if (a_seg['a_arr'] !== null){
                this.mark_inbound(scan_datetime, 'UNEXPECTED_INBOUND', true, false)
                this.make_new(null, location, a_seg['idx'])
            } else {
                this.mark_inbound(
                    scan_datetime, 'LOCATION_MISMATCH', true, false)
            }
        }
        return false
    },

    parse_outbound: function(location, connection, scan_datetime){
        if (this.active == null) {
            console.error('ERROR (parse_outbound): No active nodes found')
            return
        }

        a_seg = this.__segments[this.active]

        // console.log(JSON.stringify(a_seg))
        // console.log('location '+ location)
        // console.log('connection '+ connection)
        // console.log('scan_datetime '+ scan_datetime)

        if (a_seg['a_arr'] === null){
            this.mark_inbound(scan_datetime, 'UNEXPECTED_OUTBOUND', true, false)
        } else if (a_seg['src'] === location){
            if (a_seg['conn'] === connection){
                if (a_seg['p_dep'] >= scan_datetime){
                    this.mark_outbound(scan_datetime, undefined, false, true)
                    return true
                }
                else if ((a_seg['p_dep'] - scan_datetime) < 24 * 3600){
                    this.mark_outbound(
                        scan_datetime, 'WARN_LATE_DEPARTURE', false, true)
                    return true
                }
                else{
                    this.mark_outbound(
                        scan_datetime, 'CONNECTION_MISMATCH', true, false)
                }
            } else {
                this.mark_outbound(
                    scan_datetime, 'CONNECTION_MISMATCH', true, false)
            }
        }
        else {
            this.mark_outbound(
                scan_datetime, 'LOCATION_MISMATCH', true, false)
            this.make_new(null, location, a_seg['idx'])
        }
        return false
    },

    lookup: function(kwargs){
        idx = null
        this.__segments.forEach(function(segment, index){
            if (Utils.match(segment, kwargs)){
                idx = index
            }
        })
        return idx
    },

    value: function(){
        return this.__segments
    }
})



exports.Parser = Parser;

File: expresso/pyexpresso/pyexpresso/manager/parser.py
'''
This class exposes parser and its utilities to update a recommended path
'''
from .utils import match


class Parser(object):
    '''
    Class representing a path with utilities to parse through it
    '''
    __segments = []
    active = None

    # __map_state_segments = {}

    def __init__(self):
        '''
        Initlialize Parser.
        Optionally add segments from existing data
        '''
        self.active = None
        self.__segments = []
        self.__lcost = 0

    def deactivate(self):
        '''
        Deactivate all future edges
        '''
        for idx, segment in enumerate(self.__segments):

            if segment['st'] == 'FUTURE':
                self.__segments[idx]['st'] = 'INACT'

    @property
    def arrival(self):
        '''
        Get the actual arrival time at active vertex
        '''
        return self.__segments[self.active]['a_arr']

    @arrival.setter
    def arrival(self, sdt):
        '''
        Set actual arrival at active vertex
        '''
        self.__segments[self.active]['a_arr'] = sdt

    def add_nullseg(self):
        '''
        Add a null segment
        '''
        segment = {
            'src': None,
            'dst': None,
            'conn': None,
            'p_arr': None,
            'a_arr': None,
            'p_dep': None,
            'a_dep': None,
            'cst': None,
            'st': 'REACHED',
            'rmk': [],
            'idx': 0,
            'par': None,
            'sol': None,
            'pdd': None,
        }
        self.__segments.append(segment)
        self.active = 0
        self.__lcost = self.__segments[self.active]['cst']

    def add_segment(self, subgraph=False, **kwargs):
        '''
        Add a segment to parser
        '''
        if subgraph:

            if len(self.__segments) == 0:
                self.add_nullseg()

        segment = {
            'src': kwargs['src'],
            'dst': kwargs['dst'],
            'conn': kwargs['conn'],
            'p_arr': kwargs['p_arr'],
            'p_dep': kwargs['p_dep'],
            'cst': kwargs['cst'],
            'a_arr': kwargs.get('a_arr', None),
            'a_dep': kwargs.get('a_dep', None),
            'st':  kwargs.get('st', 'FUTURE'),
            'rmk': kwargs.get('rmk', []),
        }

        for key, value in kwargs.items():
            if key not in [
                    'src', 'dst', 'conn', 'p_arr', 'p_dep', 'cst',
                    'a_arr', 'a_dep', 'st', 'rmk', 'idx', 'par']:
                segment[key] = value

        segment['idx'] = kwargs.get('idx', len(self.__segments))

        if not subgraph:
            segment['par'] = kwargs.get('par', len(self.__segments) - 1)
        else:
            segment['par'] = self.active
            segment['st'] = 'ACTIVE'

        if segment['st'] == 'ACTIVE':
            self.active = segment['idx']
            self.__lcost = segment['cst']

        self.__segments.append(segment)
        return len(self.__segments)

    def add_segments(self, segments, novi=False):
        '''
        Add multiple segments to parser
        '''
        if not isinstance(segments, list):
            raise TypeError('List expected. Got {}'.format(type(segments)))

        for segment in segments:
            self.add_segment(subgraph=novi, **segment)
            novi = False

    def mark_inbound(self, scan_datetime, rmk=None, fail=False, arrived=False):
        '''
        Mark inbound against path
        '''

        if arrived:
            self.__segments[self.active]['a_arr'] = scan_datetime

        if rmk:
            self.__segments[self.active]['rmk'].append(rmk)

        if fail:
            self.__segments[self.active]['st'] = 'FAIL'
            self.deactivate()
            self.active = self.__segments[self.active]['par']

    def mark_outbound(
            self, scan_datetime, rmk=None, fail=False, departed=False):
        '''
        Mark outbound against path
        '''
        if departed:
            self.__segments[self.active]['a_dep'] = scan_datetime

        if rmk:
            self.__segments[self.active]['rmk'].append(rmk)

        if fail:
            self.__segments[self.active]['st'] = 'FAIL'
            self.deactivate()
            self.active = self.__segments[self.active]['par']
        else:
            self.__segments[self.active]['st'] = 'REACHED'
            self.active = self.lookup(**{
                'par': self.__segments[self.active]['idx'],
                'st': 'FUTURE'
            })
            self.__segments[self.active]['st'] = 'ACTIVE'
            self.__lcost = self.__segments[self.active]['cst']

    def make_new(self, connection, intermediary, source=None):
        '''
        Make a duplicate node from old active to new intermediary
        and mark it reached
        '''
        if source is None:
            source = self.active
        segment = {}

        for key, value in self.__segments[source].items():
            segment[key] = value
        segment['conn'] = connection
        segment['dst'] = intermediary
        segment['st'] = 'REACHED'
        segment['idx'] = len(self.__segments)
        segment['cst'] = self.lcost
        self.__segments.append(segment)
        self.active = segment['idx']
        self.__lcost = segment['cst']

    def make_new_blank(self, src, dst, cid, sdt):
        '''
        Make a new node from provided parameters and mark it reached
        '''
        segment = {
            'src': src,
            'dst': dst,
            'conn': cid,
            'p_arr': None,
            'a_arr': None,
            'p_dep': None,
            'a_dep': sdt,
            'cst': self.__lcost,
            'st': 'REACHED',
            'rmk': [],
            'par': self.active,
            'sol': self.__segments[self.active]['sol'],
            'pdd': self.__segments[self.active]['pdd'],
        }
        segment['idx'] = len(self.__segments)
        self.__segments.append(segment)
        self.active = segment['idx']
        self.__lcost = segment['cst']

    def mark_termination(self, msg):
        '''
        Mark termination of ExPath due to bad/missing data
        '''
        self.__segments[self.active]['rmk'].append(msg)
        self.__segments[self.active]['st'] = 'FAIL'

    def parse_inbound(self, location, scan_datetime):
        '''
        Parse inbound against graph
        [in] location: Location at which inbound has been performed
        [in] scan_datetime: Date time of inscan
        [out] boolean: True on successful inbound, False on failed inbound
        '''
        if self.active is None:
            raise ValueError('No active nodes found')
        a_seg = self.__segments[self.active]

        if a_seg['src'] == location:

            if a_seg['a_arr']:
                # Duplicate inbound. Ignore
                return True

            if a_seg['p_arr'] >= scan_datetime:
                self.mark_inbound(scan_datetime, arrived=True)
                return True
            elif scan_datetime < a_seg['p_dep']:
                self.mark_inbound(
                    scan_datetime, rmk='WARN_LATE_ARRIVAL', arrived=True)
                return True
            else:
                self.mark_inbound(
                    scan_datetime, rmk='FAIL_LATE_ARRIVAL', fail=True,
                    arrived=True)
        else:
            if a_seg['a_arr']:
                # Expecting outbound, inbound happened
                self.mark_inbound(
                    scan_datetime, rmk='UNEXPECTED_INBOUND', fail=True,
                    arrived=False)
                self.make_new(None, location, a_seg['idx'])
            else:
                self.mark_inbound(
                    scan_datetime, rmk='LOCATION_MISMATCH', fail=True,
                    arrived=False)
        return False

    def parse_outbound(self, location, connection, scan_datetime):
        '''
        Parse outbound against graph
        '''

        if self.active is None:
            raise ValueError('No active nodes found')
        a_seg = self.__segments[self.active]

        if not a_seg['a_arr']:
            self.mark_inbound(
                scan_datetime, rmk='UNEXPECTED_OUTBOUND', fail=True)

        elif a_seg['src'] == location:

            if a_seg['conn'] == connection:

                if a_seg['p_dep'] >= scan_datetime:
                    self.mark_outbound(scan_datetime, departed=True)
                    return True
                elif a_seg['p_dep'] - scan_datetime < 24 * 3600:
                    self.mark_outbound(
                        scan_datetime, rmk='WARN_LATE_DEPARTURE',
                        departed=True)
                    return True
                else:
                    self.mark_outbound(
                        scan_datetime, rmk='CONNECTION_MISMATCH', fail=True)
            else:
                self.mark_outbound(
                    scan_datetime, rmk='CONNECTION_MISMATCH', fail=True)
        else:
            self.mark_outbound(
                scan_datetime, rmk='LOCATION_MISMATCH', fail=True)
            self.make_new(None, location, a_seg['idx'])
        return False

    def lookup(self, **kwargs):
        '''
        Return the index for segment matching kwargs
        '''
        for index, segment in enumerate(self.__segments):

            if match(segment, kwargs):
                return index
        return None

    @property
    def value(self):
        '''
        Return the entire graph as a list
        '''
        return self.__segments

    @property
    def lcost(self):
        '''
        Minimum cost offset
        '''
        return self.__lcost
